core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+package zio.internal
+
+import zio.internal.ConcurrentWeakHashSet.Ref
+
+import java.lang.ref.WeakReference
+import scala.collection.mutable
+import java.util.concurrent.locks.ReentrantLock
+import java.lang.ref.ReferenceQueue
+import java.util.concurrent.atomic.AtomicInteger
+
+private[zio] object ConcurrentWeakHashSet {
+
+  private[internal] val MaxConcurrencyLevel: Int     = 1 << 16
+  private[internal] val MaxSegmentSize: Int          = 1 << 30
+  private[internal] val DefaultInitialCapacity: Int  = 16
+  private[internal] val DefaultLoadFactor: Float     = 0.75f
+  private[internal] val DefaultConcurrencyLevel: Int = 16
+
+  /**
+   * Reference node that links elements in the set with the same hash code
+   *
+   * @param hash
+   *   hash code of the element
+   * @param element
+   *   value passed to the weak reference
+   * @param queue
+   *   reference queue that will be used to track garbage collected elements
+   * @param nextRef
+   *   next reference node in the chain
+   * @tparam V
+   *   type of the element
+   */
+  protected class Ref[V](
+    val hash: Int,
```

Comment:
The decision to store versus compute `hash` should be made as follows:

 - If a hash code is used only once, or maybe a tiny number of times, relative to the life of the object, then it makes sense to compute it each time
 - If a hash code is used many times, it makes sense to store it

The int requires 4 bytes which adds to heap pressure and increases storage space.

Reply:
I've run some simple benchmarks for a `Ref` without cached hash it directly impacted the performance of matching. Such change affects:

* `findInChain` - matching node with the same value has to use `equals` that is pretty complex comparing to the comparisong of two ints
* `restructure` - each restrucuted reference (so pretty much all values in the segment) my need to be recalculated

<details>
  <summary>Fast benchmark 2 warmup / 2 iterations for 2s</summary>

```
[info] Benchmark                                                       Mode  Cnt      Score   Error   Units
[info] ConcurrentWeakHashSetAddBenchmark.javaAddConcurrent            thrpt    2   2141,663          ops/ms
[info] ConcurrentWeakHashSetAddBenchmark.javaAddSerial                thrpt    2   2785,435          ops/ms
[info] ConcurrentWeakHashSetAddBenchmark.springAddConcurrent          thrpt    2   1527,516          ops/ms
[info] ConcurrentWeakHashSetAddBenchmark.springAddSerial              thrpt    2   2101,580          ops/ms
[info] ConcurrentWeakHashSetAddBenchmark.zioAddConcurrent             thrpt    2   1510,795          ops/ms
[info] ConcurrentWeakHashSetAddBenchmark.zioAddSerial                 thrpt    2   2419,780          ops/ms
[info] ConcurrentWeakHashSetIterateBenchmark.javaIterateConcurrent    thrpt    2    121,800          ops/ms
[info] ConcurrentWeakHashSetIterateBenchmark.javaIterateSerial        thrpt    2    146,162          ops/ms
[info] ConcurrentWeakHashSetIterateBenchmark.springIterateConcurrent  thrpt    2    572,904          ops/ms
[info] ConcurrentWeakHashSetIterateBenchmark.springIterateSerial      thrpt    2    116,834          ops/ms
[info] ConcurrentWeakHashSetIterateBenchmark.zioIterateConcurrent     thrpt    2    593,624          ops/ms
[info] ConcurrentWeakHashSetIterateBenchmark.zioIterateSerial         thrpt    2    120,774          ops/ms
[info] ConcurrentWeakHashSetRemoveBenchmark.javaRemoveConcurrent      thrpt    2   5769,217          ops/ms
[info] ConcurrentWeakHashSetRemoveBenchmark.javaRemoveSerial          thrpt    2  18724,470          ops/ms
[info] ConcurrentWeakHashSetRemoveBenchmark.springRemoveConcurrent    thrpt    2    112,343          ops/ms
[info] ConcurrentWeakHashSetRemoveBenchmark.springRemoveSerial        thrpt    2     20,790          ops/ms
[info] ConcurrentWeakHashSetRemoveBenchmark.zioRemoveConcurrent       thrpt    2  13635,635          ops/ms
[info] ConcurrentWeakHashSetRemoveBenchmark.zioRemoveSerial           thrpt    2  10140,362          ops/ms
```

</details>